src/http_client_async.zig

```diff
@@ -397,9 +413,85 @@ fn NewHTTPContext(comptime ssl: bool) type {
             ) void {
+                // on SSL we need to do wait for the handshake
+                if (comptime ssl == false) {
+                    const active = ActiveSocket.from(bun.cast(**anyopaque, ptr).*);
+                    if (active.get(HTTPClient)) |client| {
+                        return client.onOpen(comptime ssl, socket);
+                    }
+
+                    if (active.get(PooledSocket)) |pooled| {
+                        std.debug.assert(context().pending_sockets.put(pooled));
+                    }
+
+                    socket.ext(**anyopaque).?.* = bun.cast(**anyopaque, ActiveSocket.init(&dead_socket).ptr());
+                    socket.close(0, null);
+                    if (comptime Environment.allow_assert) {
+                        std.debug.assert(false);
+                    }
+                } else {
+                    const active = ActiveSocket.from(bun.cast(**anyopaque, ptr).*);
+                    if (active.get(HTTPClient)) |client| {
+                        var ssl_ptr: *BoringSSL.SSL = @as(*BoringSSL.SSL, @ptrCast(socket.getNativeHandle()));
+                        if (!ssl_ptr.isInitFinished()) {
+                            var _hostname = client.hostname orelse client.url.hostname;
+                            if (client.http_proxy) |proxy| {
+                                _hostname = proxy.hostname;
+                            }
+
+                            var hostname: [:0]const u8 = "";
+                            var hostname_needs_free = false;
+                            if (!strings.isIPAddress(_hostname)) {
+                                if (_hostname.len < temp_hostname.len) {
+                                    @memcpy(temp_hostname[0.._hostname.len], _hostname);
+                                    temp_hostname[_hostname.len] = 0;
+                                    hostname = temp_hostname[0.._hostname.len :0];
+                                } else {
+                                    hostname = bun.default_allocator.dupeZ(u8, _hostname) catch unreachable;
+                                    hostname_needs_free = true;
+                                }
+                            }
+
+                            defer if (hostname_needs_free) bun.default_allocator.free(hostname);
+
+                            ssl_ptr.configureHTTPClient(hostname);
+                        }
+                    }
+                }
+            }
+            pub fn onHandshake(
+                ptr: *anyopaque,
+                socket: HTTPSocket,
+                success: i32,
+                ssl_error: uws.us_bun_verify_error_t,
+            ) void {
+                const authorized = if (success == 1) true else false;
+
+                const handshake_error = HTTPCertError{
+                    .error_no = ssl_error.error_no,
+                    .code = if (ssl_error.code == null) "" else ssl_error.code[0..bun.len(ssl_error.code) :0],
```

Comment:
is this memory static?

Reply:
this is inside BoringSSL ctx, but we cloned if we need outside `onCertError` because its the recomended